nifi-nar-bundles/nifi-social-media-bundle/nifi-twitter-processors/src/main/java/org/apache/nifi/processors/twitter/TwitterStreamAPI.java

```diff
@@ -0,0 +1,174 @@
+        this.expansions = parseCommaSeparatedListPropreties(context, EXPANSIONS_PROPERTY_NAME);
+        this.backfillMinutes = context.getProperty(BACKFILL_MINUTES_PROPERTY_NAME).asInteger();
+
+        TwitterCredentialsBearer creds = new TwitterCredentialsBearer(context.getProperty(BEARER_TOKEN_PROPERTY_NAME).getValue());
+        api = new TwitterApi();
+        api.setTwitterCredentials(creds);
+
+        this.executorService = Executors.newSingleThreadExecutor();
+    }
+
+    public String getBasePath() {
+        return api.getApiClient().getBasePath();
+    }
+
+    public void setBasePath(final String path) {
+        api.getApiClient().setBasePath(path);
+    }
+
+    /**
+     * This method would be called when we would like the stream to get started. This method will spin off a thread that
+     * will continue to queue tweets on to the given queue passed in the constructor. The thread will continue
+     * to run until {@code stop} is called.
+     * @param endpoint {@code TwitterStreamAPI.SAMPLE_ENDPOINT} or {@code TwitterStreamAPI.SEARCH_ENDPOINT}
+     */
+    public void start(final String endpoint) {
+        try {
+            if (endpoint.equals(SAMPLE_ENDPOINT)) {
+                stream = api.tweets().sampleStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+            } else if (endpoint.equals(SEARCH_ENDPOINT)) {
+                stream = api.tweets().searchStream(expansions, tweetFields, userFields, mediaFields, placeFields, pollFields, backfillMinutes);
+            } else {
+                throw new ProcessException("Endpoint was invalid value: " + endpoint);
+            }
+        } catch (ApiException e) {
+            logger.error("Received error {}: {}", e.getCode(), e.getMessage());
+        }
+
+        executorService.execute(new TweetQueuer());
+    }
+
+    /**
+     * This method would be called when we would like the stream to get stopped. The stream will be closed and the
+     * executorService will be shut down. If it fails to shutdown, then it will be forcefully terminated.
+     */
+    public void stop() {
+        try {
+            stream.close();
+        } catch (IOException e) {
+            logger.error("IOException occurred while closing stream: {}", e.getMessage());
+        }
+
+        executorService.shutdown();
+        try {
+            if (!executorService.awaitTermination(3, TimeUnit.SECONDS)) {
+                executorService.shutdownNow();
+            }
+        } catch (InterruptedException e) {
+            executorService.shutdownNow();
+            logger.error("InterruptedException occurred while waiting for executor service to shut down: {}", e.getMessage());
+        }
+    }
+
+    private class TweetQueuer implements Runnable {
+        @Override
+        public void run() {
+            try (BufferedReader reader = new BufferedReader(new InputStreamReader(stream))) {
+                String tweetLine = reader.readLine();
```

Comment:
Minor naming recommendation, adjusting the variable name and adding a comment helps communicate the fact that Twitter returns a complete JSON record in each line:
```suggestion
                // Twitter returns one JSON object in each line of the stream
                String tweetRecord = reader.readLine();
```

Reply:
Agreed, making the changes